### PR TITLE
Major changes of Protocols and Async/Await in Apexy

### PR DESCRIPTION
**Pull request by laqiguy**
- Different approaches of async code separated in different protocols
- Updated Async/Await methods
- Some code reorganisation
- macOS and Xcode images updated
- Some documentation updated
- Updated Alamofire version

### CODE REVIEW

File: Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        let result = APIResult<T.Content>(catching: { () throws -> T.Content in
+            do {
+                let data = try dataResponse.result.get()
+                return try endpoint.content(from: dataResponse.response, with: data)
+            } catch {
+                throw error.unwrapAlamofireValidationError()
+            }
+        })
+
+        Task.detached {[weak self, dataResponse, result] in
```
Comment (vani2) [inline Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift]:
```suggestion
        Task.detached { [weak self, dataResponse, result] in
```

File: Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import Apexy
+import Foundation
+
+@available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, *)
+extension AlamofireClient: ConcurrencyClient {
+    
+    func observeResponse(
+        dataResponse: DataResponse<Data, AFError>,
+        error: Error?) async {
+            return await withCheckedContinuation{ continuation in
```
Comment (vani2) [inline Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift]:
```suggestion
            await withCheckedContinuation{ continuation in
```

File: Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        let result = APIResult<T.Content>(catching: { () throws -> T.Content in
+            do {
+                let data = try dataResponse.result.get()
+                return try endpoint.content(from: dataResponse.response, with: data)
+            } catch {
+                throw error.unwrapAlamofireValidationError()
+            }
+        })
+
+        Task.detached {[weak self, dataResponse, result] in
```
Comment (vani2) [inline Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift]:
```suggestion
        Task.detached { [weak self, dataResponse, result] in
```

File: Sources/ApexyURLSession/URLSessionClient+Concurrency.swift [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import Apexy
+import Foundation
+
+@available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, *)
+extension URLSessionClient: ConcurrencyClient {
+        
+    func observeResponse(
+        request: URLRequest?,
+        responseResult: Result<(data: Data, response: URLResponse), Error>) async {
+            return await withCheckedContinuation{ continuation in
```
Comment (vani2) [inline Sources/ApexyURLSession/URLSessionClient+Concurrency.swift]:
```suggestion
            await withCheckedContinuation { continuation in
```

File: Sources/ApexyURLSession/URLSessionClient+Concurrency.swift [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            if let httpResponse = response.response as? HTTPURLResponse {
+                try endpoint.validate(request, response: httpResponse, data: response.data)
+            }
+            
+            responseResult = .success(response)
+        } catch let someError {
+            responseResult = .failure(someError)
+        }
+                
+        Task.detached {[weak self, request, responseResult] in
```
Comment (vani2) [inline Sources/ApexyURLSession/URLSessionClient+Concurrency.swift]:
```suggestion
        Task.detached { [weak self, request, responseResult] in
```

File: Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift (hunk 1 above)
Comment (vani2) [inline Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift]:
```suggestion
        Task.detached { [weak self, dataResponse, result] in
```

File: Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift (hunk 2 above)
Comment (vani2) [reply Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift]: I believe that should work

**Review by vani2 — COMMENTED** [review]
Looks good 👍. Added some minor style suggestions.

**Comment by vani2** [comment]
Don't forget to bump version and I noticed [old e-mail](https://github.com/RedMadRobot/apexy-ios/blob/02e3454d886eb910371ffead7e3e8a60af74009e/Apexy.podspec#L7) @Alexander-Ignition

**Comment by Alexander-Ignition** [comment]
> Don't forget to bump version and I noticed [old e-mail](https://github.com/RedMadRobot/apexy-ios/blob/02e3454d886eb910371ffead7e3e8a60af74009e/Apexy.podspec#L7) @Alexander-Ignition

I have transferred the rights to publish new versions to @laqiguy

File: Sources/Apexy/Clients/ConcurrencyClient.swift [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    }
+}
+
+@available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, *)
+public enum AsyncAwaitHelper {
+    public typealias ContentContinuation<T> = CheckedContinuation<T, Error>
+
+    public static func adaptToAsync<T>(dataTaskClosure: (ContentContinuation<T>) -> Progress) async throws -> T {
+        let progressWrapper = ProgressWrapper()
+        return try await withTaskCancellationHandler(handler: {
```
Comment (Alexander-Ignition) [inline Sources/Apexy/Clients/ConcurrencyClient.swift]: If the request was canceled before the start, `handler` will be called first and then `operation`. In this case, the request will still be launched, even if the cancellation was made first

A small example with cancellation

```
@main
struct App {
    static func main() async {
        do {
            let task = Task<String, Error>(priority: .background) {
                try await fetch()
            }
            task.cancel()
            let value = try await task.value
            print(value)
        } catch {
            print(error)
        }
    }

    static func fetch() async throws -> String {
        await withTaskCancellationHandler {
            print("start")
            return "success"
        } onCancel: {
            print("cancel")
        }

    }
}
```

File: Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift (hunk 1 above)
Comment (Alexander-Ignition) [inline Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift]: I didn't understand why to do a `Task.detach` that calls a method inside which `completionQueue.async` occurs.

Thus, there are two switches to other threads

File: Sources/Apexy/Clients/ConcurrencyClient.swift [hunk 7]
```diff
@@ -41,14 +41,20 @@ public extension Client where Self: ConcurrencyClient {
...
+    
     public typealias ContentContinuation<T> = CheckedContinuation<T, Error>
 
     public static func adaptToAsync<T>(dataTaskClosure: (ContentContinuation<T>) -> Progress) async throws -> T {
         let progressWrapper = ProgressWrapper()
         return try await withTaskCancellationHandler(handler: {
             progressWrapper.cancel()
         }, operation: {
-            try await withCheckedThrowingContinuation { (continuation: ContentContinuation<T>) in
+            guard progressWrapper.isCancelled else { throw AsyncError.cancelledBeforeStart }
```
Comment (Alexander-Ignition) [inline Sources/Apexy/Clients/ConcurrencyClient.swift]: `Task` has a [checkCancellation()](https://developer.apple.com/documentation/swift/task/checkcancellation()) method

File: Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            do {
+                let data = try dataResponse.result.get()
+                return try endpoint.content(from: dataResponse.response, with: data)
+            } catch {
+                throw error.unwrapAlamofireValidationError()
+            }
+        })
+
+        Task.detached { [weak self, dataResponse, result] in
+            await self?.observeResponse(dataResponse: dataResponse, error: result.error)
```
Comment (Alexander-Ignition) [inline Sources/ApexyAlamofire/AlamofireClient+Concurrency.swift]: Synchronous code can be executed in `Task` without `withCheckedContinuation`

```swift
Task.detached { [weak self, dataResponse, result] in
    self?.observeResponse(dataResponse: dataResponse, error: result.error)
}
```

File: Tests/ApexyURLSessionTests/AsyncAwaitHelperTests.swift [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        let task = Task<String, Error>(priority: .background) {
+            try await AsyncAwaitHelper.adaptToAsync(dataTaskClosure: { continuation in
+                continuation.resume(returning: "123")
+                return Progress()
+            })
+        }
+        task.cancel()
+        do {
+            _ = try await task.value
+            XCTAssert(false)
```
Comment (Alexander-Ignition) [inline Tests/ApexyURLSessionTests/AsyncAwaitHelperTests.swift]: XCTest has the [XCTFail](https://developer.apple.com/documentation/xctest/1500970-xctfail) function

**Comment by laqiguy** [comment]
Finally I fixed comments. Can you review?
@Alexander-Ignition @modestman

**Review by Alexander-Ignition — APPROVED** [review]

**Review by modestman — APPROVED** [review]

**Pull request merged**